### DISCUSSION TITLE
[FAIL: invariant_protocolMustHaveMoreValueThanTotalSupply replay failure]

### DISCUSSION THREAD

**Discussion by muhammadjehanzaib** [Cyfrin/foundry-full-course-cu#2938] [General]
Reverting on First call, because weth and wbtc balance isn't updating and assert triggered(         assert(wbtcValue + wethValue >= totalSupply);
).

forge test --mt invariant_protocolMustHaveMoreValueThanTotalSupply -vvvv
[⠒] Compiling...
No files changed, compilation skipped

Ran 1 test for test/fuzz/InvariantsTest.t.sol:InvariantTest
[FAIL: invariant_protocolMustHaveMoreValueThanTotalSupply replay failure]
        [Sequence]
                sender=0x0000000000000000175f474921Aa6e14b39E8fdA addr=[test/fuzz/Handler.t.sol:Handler]0x2e234DAe75C793f67A35089C9d99245E1C58470b calldata=depositCollateral(uint256,uint256) args=[245, 145806344889484429280554321950436350630990 [1.458e41]]
                sender=0x0000000000000000175f474921Aa6e14b39E8fdA addr=[test/fuzz/Handler.t.sol:Handler]0x2e234DAe75C793f67A35089C9d99245E1C58470b calldata=mintDSC(uint256,uint256) args=[20923119790302747841292182215721325678979221970086581278732577534 [2.092e64], 12145472341703325850398765862 [1.214e28]]
 invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 1, calls: 1, reverts: 1)
Logs:
  Bound result 61145621309733733021492620885
  Bound result 20120506031947133867094539127700
  weth value 0
  wbtc value 0
  total supply 20120506031947133867094539127700
  Times mint called 1
  Times Deposit Callateral called 2

Traces:
  [11165006] InvariantTest::setUp()
    ├─ [4343714] → new DeployDSC@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 21583 bytes of code
    ├─ [5096743] DeployDSC::run()
    │   ├─ [3299777] → new <unknown>@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [372255] → new <unknown>@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   └─ ← [Return] 1082 bytes of code
    │   │   ├─ [516505] → new <unknown>@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DeployDSC: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], value: 2000000000000000000000 [2e21])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [372255] → new <unknown>@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   │   └─ ← [Return] 1082 bytes of code
    │   │   ├─ [516505] → new <unknown>@0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DeployDSC: [0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f], value: 1000000000000000000000 [1e21])
    │   │   │   └─ ← [Return] 2117 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 6163 bytes of code
    │   ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    │   ├─ [0] VM::startBroadcast(<pk>)
    │   │   └─ ← [Return] 
    │   ├─ [591255] → new <unknown>@0x5FbDB2315678afecb367f032d93F642f64180aa3
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0x0000000000000000000000000000000000000000, newOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266)
    │   │   └─ ← [Return] 2609 bytes of code
    │   ├─ [920716] → new DSCEngine@0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512
    │   │   └─ ← [Return] 3921 bytes of code
    │   ├─ [2424] 0x5FbDB2315678afecb367f032d93F642f64180aa3::transferOwnership(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   ├─ emit OwnershipTransferred(previousOwner: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266, newOwner: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512])
    │   │   └─ ← [Stop] 
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   └─ ← [Return] 0x5FbDB2315678afecb367f032d93F642f64180aa3, DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 0x104fBc016F4bb334D775a19E8A6510109AC63E00
    ├─ [912] 0x104fBc016F4bb334D775a19E8A6510109AC63E00::activeNetworkConfig() [staticcall]
    │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1, 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3, 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 77814517325470205911140941194401928579557062014761831930645393041380819009408 [7.781e76]
    ├─ [1469611] → new Handler@0x2e234DAe75C793f67A35089C9d99245E1C58470b
    │   ├─ [1312] DSCEngine::getCollateralTokens() [staticcall]
    │   │   └─ ← [Return] [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76]
    │   ├─ [580] DSCEngine::getCollateralTokenPriceFeed(0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496) [staticcall]
    │   │   └─ ← [Return] 0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   ├─ [580] DSCEngine::getCollateralTokenPriceFeed(0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76) [staticcall]
    │   │   └─ ← [Return] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   └─ ← [Return] 6538 bytes of code
    └─ ← [Stop] 

  [237329] Handler::depositCollateral(245, 145806344889484429280554321950436350630990 [1.458e41])
    ├─ [0] console::log("Bound result", 61145621309733733021492620885 [6.114e28]) [staticcall]
    │   └─ ← [Stop] 
    ├─ [0] VM::startPrank(0x0000000000000000175f474921Aa6e14b39E8fdA)
    │   └─ ← [Return] 
    ├─ [29712] 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76::mint(0x0000000000000000175f474921Aa6e14b39E8fdA, 61145621309733733021492620885 [6.114e28])
    │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x0000000000000000175f474921Aa6e14b39E8fdA, value: 61145621309733733021492620885 [6.114e28])
    │   └─ ← [Stop] 
    ├─ [24739] 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76::approve(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 61145621309733733021492620885 [6.114e28])
    │   ├─ emit Approval(owner: 0x0000000000000000175f474921Aa6e14b39E8fdA, spender: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 61145621309733733021492620885 [6.114e28])
    │   └─ ← [Return] true
    ├─ [98062] DSCEngine::depositCollateral(0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 61145621309733733021492620885 [6.114e28])
    │   ├─ emit CollateralDeposited(user: 0x0000000000000000175f474921Aa6e14b39E8fdA, token: 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, amount: 61145621309733733021492620885 [6.114e28])
    │   ├─ [26088] 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76::transferFrom(0x0000000000000000175f474921Aa6e14b39E8fdA, DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], 61145621309733733021492620885 [6.114e28])
    │   │   ├─ emit Transfer(from: 0x0000000000000000175f474921Aa6e14b39E8fdA, to: DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512], value: 61145621309733733021492620885 [6.114e28])
    │   │   └─ ← [Return] true
    │   ├─ [8993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   ├─ [8993] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 100000000000 [1e11], 1, 1, 1
    │   └─ ← [Stop] 
    ├─ [0] VM::stopPrank()
    │   └─ ← [Return] 
    └─ ← [Stop] 

  [168489] Handler::mintDSC(20923119790302747841292182215721325678979221970086581278732577534 [2.092e64], 12145472341703325850398765862 [1.214e28])
    ├─ [43725] DSCEngine::getAccontInformation(0x0000000000000000175f474921Aa6e14b39E8fdA) [staticcall]
    │   ├─ [8993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   ├─ [8993] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 100000000000 [1e11], 1, 1, 1
    │   └─ ← [Return] 0, 61145621309733733021492620885000 [6.114e31]
    ├─ [0] console::log("Bound result", 20120506031947133867094539127700 [2.012e31]) [staticcall]
    │   └─ ← [Stop] 
    ├─ [0] VM::startPrank(0x0000000000000000175f474921Aa6e14b39E8fdA)
    │   └─ ← [Return] 
    ├─ [84524] DSCEngine::mintDsc(20120506031947133867094539127700 [2.012e31])
    │   ├─ [993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   ├─ [993] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 100000000000 [1e11], 1, 1, 1
    │   ├─ [49149] 0x5FbDB2315678afecb367f032d93F642f64180aa3::mint(0x0000000000000000175f474921Aa6e14b39E8fdA, 20120506031947133867094539127700 [2.012e31])
    │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: 0x0000000000000000175f474921Aa6e14b39E8fdA, value: 20120506031947133867094539127700 [2.012e31])
    │   │   └─ ← [Return] true
    │   └─ ← [Stop] 
    ├─ [0] VM::stopPrank()
    │   └─ ← [Return] 
    └─ ← [Stop] 

  [75094] InvariantTest::invariant_protocolMustHaveMoreValueThanTotalSupply()
    ├─ [2349] 0x5FbDB2315678afecb367f032d93F642f64180aa3::totalSupply() [staticcall]
    │   └─ ← [Return] 20120506031947133867094539127700 [2.012e31]
    ├─ [2607] 0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496::balanceOf(DSCEngine: [0xe7f1725E7734CE288F8367e1Bb143E90bb3F0512]) [staticcall]
    │   └─ ← [Return] 0
    ├─ [2607] 0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76::balanceOf(0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76) [staticcall]
    │   └─ ← [Return] 0
    ├─ [15036] DSCEngine::getUsdValue(0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496, 0) [staticcall]
    │   ├─ [8993] 0x90193C961A926261B756D1E5bb255e67ff9498A1::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 200000000000 [2e11], 1, 1, 1
    │   └─ ← [Return] 0
    ├─ [15036] DSCEngine::getUsdValue(0xDB8cFf278adCCF9E9b5da745B44E754fC4EE3C76, 0) [staticcall]
    │   ├─ [8993] 0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3::latestRoundData() [staticcall]
    │   │   └─ ← [Return] 1, 100000000000 [1e11], 1, 1, 1
    │   └─ ← [Return] 0
    ├─ [0] console::log("weth value", 0) [staticcall]
    │   └─ ← [Stop] 
    ├─ [0] console::log("wbtc value", 0) [staticcall]
    │   └─ ← [Stop] 
    ├─ [0] console::log("total supply", 20120506031947133867094539127700 [2.012e31]) [staticcall]
    │   └─ ← [Stop] 
    ├─ [2362] Handler::timesMintIsCalled() [staticcall]
    │   └─ ← [Return] 1
    ├─ [0] console::log("Times mint called", 1) [staticcall]
    │   └─ ← [Stop] 
    ├─ [2405] Handler::timeDepositCollateralCalled() [staticcall]
    │   └─ ← [Return] 2
    ├─ [0] console::log("Times Deposit Callateral called", 2) [staticcall]
    │   └─ ← [Stop] 
    └─ ← [Revert] panic: assertion failed (0x01)

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 5.31ms (2.28ms CPU time)

Ran 1 test suite in 1.87s (5.31ms CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/fuzz/InvariantsTest.t.sol:InvariantTest
[FAIL: invariant_protocolMustHaveMoreValueThanTotalSupply replay failure]
        [Sequence]
                sender=0x0000000000000000175f474921Aa6e14b39E8fdA addr=[test/fuzz/Handler.t.sol:Handler]0x2e234DAe75C793f67A35089C9d99245E1C58470b calldata=depositCollateral(uint256,uint256) args=[245, 145806344889484429280554321950436350630990 [1.458e41]]
                sender=0x0000000000000000175f474921Aa6e14b39E8fdA addr=[test/fuzz/Handler.t.sol:Handler]0x2e234DAe75C793f67A35089C9d99245E1C58470b calldata=mintDSC(uint256,uint256) args=[20923119790302747841292182215721325678979221970086581278732577534 [2.092e64], 12145472341703325850398765862 [1.214e28]]
 invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 1, calls: 1, reverts: 1)

Encountered a total of 1 failing tests, 0 tests succeeded

**Comment by muhammadjehanzaib** [comment]
InvariantsTest.t.sol

//SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {Test, console} from "forge-std/Test.sol";
import {StdInvariant} from "forge-std/StdInvariant.sol";
import {DSCEngine} from "../../src/DSCEngine.sol";
import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";
import {DeployDSC} from "../../script/DeployDSC.s.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import {Handler} from "./Handler.t.sol";

contract InvariantTest is StdInvariant, Test {
    DSCEngine public dscEngine;
    DecentralizedStableCoin public dsc;
    HelperConfig public config;
    DeployDSC public deploy;
    address weth;
    address wbtc;
    Handler handler;

    function setUp() external {
        deploy = new DeployDSC();
        (dsc, dscEngine, config) = deploy.run();
        (,, weth, wbtc,) = config.activeNetworkConfig();
        // targetContract(address(dscEngine));
        handler = new Handler(dscEngine, dsc);
        targetContract(address(handler));
    }

    function invariant_protocolMustHaveMoreValueThanTotalSupply() public view {
        uint256 totalSupply = dsc.totalSupply();
        uint256 totalWethDeposited = IERC20(weth).balanceOf(address(dscEngine));
        uint256 totalBtcDeposited = IERC20(wbtc).balanceOf(address(wbtc));

        uint256 wethValue = dscEngine.getUsdValue(weth, totalWethDeposited);
        uint256 wbtcValue = dscEngine.getUsdValue(wbtc, totalBtcDeposited);

        console.log("weth value", wethValue);
        console.log("wbtc value", wbtcValue);
        console.log("total supply", totalSupply);
        console.log("Times mint called", handler.timesMintIsCalled());
        console.log("Times Deposit Callateral called", handler.timeDepositCollateralCalled());

        assert(wbtcValue + wethValue >= totalSupply);
    }
}



Handler.t.sol

//SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import {Test} from "forge-std/Test.sol";
import {DSCEngine} from "../../src/DSCEngine.sol";
import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
import {ERC20Mock} from "../Mocks/ERC20Mock.sol";
import {MockV3Aggregator} from "../Mocks/MockV3Aggregator.sol";

contract Handler is Test {
    DSCEngine dsce;
    DecentralizedStableCoin dsc;

    MockV3Aggregator public ethUsdPriceFeed;
    MockV3Aggregator public btcUsdPriceFeed;

    ERC20Mock public weth;
    ERC20Mock public wbtc;

    address[] public userWithCollateralDeposited;
    uint256 public timesMintIsCalled;
    uint256 public timeDepositCollateralCalled;

    constructor(DSCEngine _engine, DecentralizedStableCoin _dsc) {
        dsce = _engine;
        dsc = _dsc;

        address[] memory collateralTokens = dsce.getCollateralTokens();
        weth = ERC20Mock(collateralTokens[0]);
        wbtc = ERC20Mock(collateralTokens[1]);

        ethUsdPriceFeed = MockV3Aggregator(dsce.getCollateralTokenPriceFeed(address(weth)));
        btcUsdPriceFeed = MockV3Aggregator(dsce.getCollateralTokenPriceFeed(address(wbtc)));
    }

    function mintDSC(uint256 amount, uint256 addressSeed) public {
        if(userWithCollateralDeposited.length == 0) {return;}
        address sender = userWithCollateralDeposited[addressSeed % userWithCollateralDeposited.length];
        (uint256 totalDSCMinted, uint256 collateralValueInUsd) = dsce.getAccontInformation(msg.sender);
        int256 maxDSCToMint = ((int256(collateralValueInUsd)) / 2) - int256(totalDSCMinted);
        timesMintIsCalled++;
        if(maxDSCToMint < 0) {
            return;
        }
        amount = bound(amount, 0, uint256(maxDSCToMint));
        vm.startPrank(sender);
        dsce.mintDsc(amount);
        vm.stopPrank();
    }

    function depositCollateral(uint256 collateralSeed, uint256 amountCollateral) public {
        amountCollateral = bound(amountCollateral, 1, type(uint96).max);
        ERC20Mock collateral = _getCollateralSFromSeed(collateralSeed);

        vm.startPrank(msg.sender);
        collateral.mint(msg.sender, amountCollateral);
        collateral.approve(address(dsce), amountCollateral);
        dsce.depositCollateral(address(collateral), amountCollateral);
        vm.stopPrank();
        userWithCollateralDeposited.push(msg.sender);
        timeDepositCollateralCalled +=2;
    }

    function redeemCollaterl(uint256 collateralSeed, uint256 amountCollateral) public {
        ERC20Mock collateral = _getCollateralSFromSeed(collateralSeed);
        uint256 maxCollateralToRedeem = dsce.getCollateralBalanceOfUser(msg.sender, address(collateral));

        amountCollateral = bound(amountCollateral, 0, maxCollateralToRedeem);
        if (amountCollateral == 0) {
            return;
        }
        vm.prank(msg.sender);
        dsce.redeemCollateral(address(collateral), amountCollateral);
    }

    function _getCollateralSFromSeed(uint256 collateralSeed) private view returns (ERC20Mock) {
        if (collateralSeed % 2 == 0) {
            return weth;
        }
        return wbtc;
    }
}

foundry.toml

[profile.default]
src = "src"
out = "out"
libs = ["lib"]
remappings = ["@chainlink/contracts/=lib/chainlink-brownie-contracts/contracts","@openzeppelin/contracts=lib/openzeppelin-contracts/contracts"]

[invariant]
runs = 128
depth = 128
fail_on_revert = false

**Comment by muhammadjehanzaib** [comment]
Problem solved.

Invariant Updated
function invariant_protocolMustHaveMoreValueThanTotalSupply() public view {
        uint256 totalSupply = dsc.totalSupply();
        uint256 totalWethDeposited = IERC20(weth).balanceOf(address(dscEngine));
        uint256 totalBtcDeposited = IERC20(wbtc).balanceOf(address(dscEngine));

        uint256 wethValue = dscEngine.getUsdValue(weth, totalWethDeposited);
        uint256 wbtcValue = dscEngine.getUsdValue(wbtc, totalBtcDeposited);

        console.log("weth value", wethValue);
        console.log("wbtc value", wbtcValue);
        console.log("total supply", totalSupply);
        console.log("Times mint called", handler.timesMintIsCalled());
        console.log("Times Deposit Callateral called", handler.timeDepositCollateralCalled());

        assert(wbtcValue + wethValue >= totalSupply);
    }

**Reply by EngrPips** [reply]
I am glad you figured this out.
